In [1]:
from datasets import Dataset

base_train = Dataset.load_from_disk('/data/ephemeral/home/sungeun/level2-mrc-nlp-13/data/train_dataset/train')
base_train_df = base_train.to_pandas()
base_train_df.head()

/opt/conda/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


title                                            context  \
0    미국 상원  미국 상의원 또는 미국 상원(United States Senate)은 양원제인 미국...   
1   인사조직관리  '근대적 경영학' 또는 '고전적 경영학'에서 현대적 경영학으로 전환되는 시기는 19...   
2      강희제  강희제는 강화된 황권으로 거의 황제 중심의 독단적으로 나라를 이끌어 갔기에 자칫 전...   
3   금동삼존불감  불상을 모시기 위해 나무나 돌, 쇠 등을 깎아 일반적인 건축물보다 작은 규모로 만든...   
4  계사명 사리구  동아대학교박물관에서 소장하고 있는 계사명 사리구는 총 4개의 용기로 구성된 조선후기...   

                                  question            id  \
0         대통령을 포함한 미국의 행정부 견제권을 갖는 국가 기관은?  mrc-1-000067   
1                   현대적 인사조직관리의 시발점이 된 책은?  mrc-0-004397   
2           강희제가 1717년에 쓴 글은 누구를 위해 쓰여졌는가?  mrc-1-000362   
3  11~12세기에 제작된 본존불은 보통 어떤 나라의 특징이 전파되었나요?  mrc-0-001510   
4               명문이 적힌 유물을 구성하는 그릇의 총 개수는?  mrc-0-000823   

                                         answers  document_id  \
0        {'answer_start': [235], 'text': ['하원']}        18293   
1  {'answer_start': [212], 'text': ['《경영의 실제》']}        51638   
2        {'answer_start': [510], 'text': ['백성']}         5028   
3        {'answer_start': [625], 'text': ['중국']}        34146   
4         {'answer_start': [30], 'text': ['4개']}        47334   

   __index_level_0__  
0                 42  
1               2873  
2                230  
3                992  
4                548

In [2]:
base_train_df.iloc[0]

title                                                            미국 상원
context              미국 상의원 또는 미국 상원(United States Senate)은 양원제인 미국...
question                              대통령을 포함한 미국의 행정부 견제권을 갖는 국가 기관은?
id                                                        mrc-1-000067
answers                        {'answer_start': [235], 'text': ['하원']}
document_id                                                      18293
__index_level_0__                                                   42
Name: 0, dtype: object

In [3]:
train_most_common_title = base_train_df['title'].value_counts().head(10)
print(train_most_common_title)

title
윤치호            43
장면             24
여운형            23
루트비히 비트겐슈타인    13
데카메론의 줄거리      13
근대             13
강희제            11
이오 (위성)        11
미치류            11
자유한국당          10
Name: count, dtype: int64


In [4]:
# 'title'이 '윤치호'인 행을 먼저 필터링한 후 중복된 'context' 제거
yunchiho_unique_context_rows = base_train_df[base_train_df['title'] == '윤치호']

# 결과 출력
yunchiho_unique_context_rows.head()

title                                            context  \
19    윤치호  혹자는 윤치호의 직계선조인 윤근수, 윤두수로 그의 문중은 당색으로는 서인 소론 계열...   
171   윤치호  1903년 1월 함경도 안핵사로 임명되어 함경남도 함흥에 파견되었고, 7월 천안군수...   
201   윤치호  그러나 한편으로는 미국 사회의 인종차별주의적, 흑인을 차별하는 태도를 목격하면서 백...   
294   윤치호  서재필이 배재학당의 젊은 학생들과 애국적인 시민을 독립협회로 모으는 데 기여했다면 ...   
415   윤치호  1895년 귀국 직후부터 그는 영어 실력을 바탕으로 그는 조선을 방문한 선교사들의 ...   

                                            question            id  \
19                        윤치호가 극찬했던 조선시대의 인물은 누구인가요?  mrc-1-000027   
171                 윤치호가 중앙 정계에서 다시 활동하기 전에 일어난 전투는?  mrc-0-000606   
201  평상시 기독교선교사들로부터 차별적 대우를 받은 윤치호가 생각한 미국의 중대한 속성은?  mrc-0-001069   
294               서재필은 어디에서 공부하는 학생들을 독립협회로 모이게 했나요?  mrc-0-000980   
415   윤치호가 아버지 윤웅렬을 설득해 교회와 학교를 지을 땅을 제공했던 선교사의 이름은?  mrc-0-004643   

                                           answers  document_id  \
19     {'answer_start': [585], 'text': ['다산 정약용']}         5322   
171      {'answer_start': [616], 'text': ['러일전쟁']}         5295   
201      {'answer_start': [645], 'text': ['인종주의']}         5284   
294        {'answer_start': [5], 'text': ['배재학당']}         5342   
415  {'answer_start': [77], 'text': ['조세핀 애턴 캠밸']}         5289   

     __index_level_0__  
19                  18  
171                410  
201                705  
294                648  
415               3028

In [5]:
len(yunchiho_unique_context_rows)

43

In [6]:
grouped = yunchiho_unique_context_rows.groupby('context')['question'].apply(list).reset_index()
grouped.head()

context  \
0  1884년 1월 18일부터 8월 9일까지 윤치호는 거듭하여 사관학교 설립을 상주한다...   
1  1888년 중국 짱수 성 상하이에서 일하던 미국 남감리교회 선교사 알렌의 주선으로 ...   
2  1895년 12월 독립협회의 동지 서재필이 복권되어 귀국했다. 서재필이 처음 귀국했...   
3  1895년 귀국 직후부터 그는 영어 실력을 바탕으로 그는 조선을 방문한 선교사들의 ...   
4  1903년 1월 함경도 안핵사로 임명되어 함경남도 함흥에 파견되었고, 7월 천안군수...   

                                            question  
0  [윤치호와 함께 개화당의 거사에 대해 비판했던 사람은?, 윤치호가 국방군 양성을 위...  
1                    [김옥균이 히라노초를 갔을 당시 믿고 의지했던 인물은?]  
2                             [1895년에 서재필이 숨어있던 곳은?]  
3  [윤치호가 아버지 윤웅렬을 설득해 교회와 학교를 지을 땅을 제공했던 선교사의 이름은...  
4                 [윤치호가 중앙 정계에서 다시 활동하기 전에 일어난 전투는?]

In [7]:
context = grouped['context'][0]
context

'1884년 1월 18일부터 8월 9일까지 윤치호는 거듭하여 사관학교 설립을 상주한다. 윤치호는 군대 통솔권의 일원화 군인정신의 합일, 상무정신의 강화를 통하여 충성스럽고 용감한 국방군을 양성해야 한다고 보았던 것이다. 이러한 목적에서 그는 미국인 군사교관을 초빙하여 각 영을 통합훈련할 것 과 사관학교 설립을 건의했던 것이다. 이어 병원과 학교의 설립 및 전신국의 설치를 미국인에게 허가해줄 것을 건의하는 등 근대시설의 도입에도 깊은 관심을 보였다\\n\\n1884년 7월에는 선교사들을 통해 신식 병원과 전화국을 유치, 개설할 것을 고종에게 상주하여 허락받았다. 그러나 신식 병원 도입과 전화국 개통은 갑신정변의 실패로 전면 백지화된다. 1894년 9월 무렵 그는 일본의 조선 침략을 예상하였다. \'일본은 이제까지는 개혁을 조선인 스스로 하도록 하려 했다. 그러나 그들이 볼 때 조선인들이 개혁의 의욕도 능력도 없음을 보고 주도권을 잡기로 결심한 것 같다 \'며 일본이 한국에 영향력을 행사하리라고 전망했다. 1884년 12월의 갑신정변 직전까지 그는 온건파 개화당의 일원으로 자주독립과 참정권, 부국강병을 위해 활동하였다. 영어 실력의 부족함을 느낀 그는 다시 주조선미국 공사관의 직원들과 교류하며 자신에게 영어를 가르쳐줄 것을 부탁하여 주조선미국공사관 직원 미군 중위 존 B. 베르나든(John B. Bernadon)이 이를 수락하였다. 5월 그는 1개월간 베르나든에게 하루 한 시간씩 영어 개인 지도를 받기도 했다.\\n\\n1884년 12월 갑신정변 초기에 윤치호는 정변 계획을 접하고 혁명의 성공을 기대하였다. 당시 김옥균을 믿고 따랐던 그는 1894년 11월에 접어들면서 윤치호는 아버지인 윤웅렬과 함께 \'개화당의 급진이 불가능한 일\'이라는 입장을 견지하고 개화당의 급진성을 겨냥, 근신을 촉구하는 입장을 보였다. 며칠 뒤 윤치호는 김옥균에게 "가친(아버지)이 기회를 보고, 변화를 엿보아 움직이는 것이 좋겠다고 한다 "라는 말을 전했다.\\n\\n그는 서광범, 김옥

In [8]:
question = grouped['question'][0]
question

['윤치호와 함께 개화당의 거사에 대해 비판했던 사람은?', '윤치호가 국방군 양성을 위해 만들자고 건의한 것은?']

- 윤치호가 영어를 배우기 위해 도움을 청한 사람은 누구였는가?
- 윤치호는 1884년 갑신정변이 실패할 것이라고 예감한 이유는 무엇인가?
- 윤치호가 미국인 군사 교관을 초빙하려 했던 목적은 무엇인가?
- 갑신정변 실패 이후 윤치호가 느낀 신변의 위협은 무엇으로부터 비롯되었는가?
- 윤치호가 1894년에 일본의 조선 침략을 예상한 이유는 무엇인가?
- 갑신정변 당시 윤치호가 개화당의 급진성에 대해 취했던 입장은 무엇이었는가?

In [16]:
pip install transformers torch

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
Note: you may need to restart the kernel to use updated packages.


In [17]:
from transformers import T5TokenizerFast, T5ForConditionalGeneration

# Load the pre-trained model and tokenizer
model_name = 'paust/pko-t5-base'  # You can also use a Korean-specific model if needed
tokenizer = T5TokenizerFast.from_pretrained(model_name)
model = T5ForConditionalGeneration.from_pretrained(model_name)

Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.


In [18]:
def generate_question(text):
    # Prepare the input text
    input_text = "generate question: " + text
    input_ids = tokenizer.encode(input_text, return_tensors='pt')

    # Generate questions
    output = model.generate(input_ids, max_length=50, num_beams=5, early_stopping=True)
    question = tokenizer.decode(output[0], skip_special_tokens=True)

    return question

# Example usage
text = "The capital of France is Paris."
question = generate_question(text)
print("Generated Question:", question)

Generated Question:  question question: The capital of France is Paris. question  The cap


In [19]:
from transformers import pipeline

# Load the Q&A pipeline
qa_pipeline = pipeline("question-answering", model="distilbert-base-uncased-distilled-squad")

def answer_question(question, context):
    result = qa_pipeline(question=question, context=context)
    return result['answer']

# Example usage
context = "The capital of France is Paris."
answer = answer_question(question, context)
print("Answer:", answer)

Answer: Paris


In [20]:
# Complete Example
text = "The capital of France is Paris."
question = generate_question(text)
answer = answer_question(question, text)

print("Generated Question:", question)
print("Answer:", answer)


Generated Question:  question question: The capital of France is Paris. question  The cap
Answer: Paris
